### PR TITLE
policy(feat): `GrpcRoute` status support

### PR DESCRIPTION
**Pull request by the-wondersmith**
## Subject

Enable handling of Gateway [`GRPCRoute`](https://gateway-api.sigs.k8s.io/api-types/grpcroute/) resources in [`linkerd-policy-controller-k8s-status`](https://github.com/linkerd/linkerd2/tree/main/policy-controller/k8s/status).

## Problem

In order to support driving traffic via Gateway `GRPCRoute` resources, the policy controller's `status` component
must be made aware of both the different route type(s) (and their associated "sources") as well as how to handle
them.

## Solution

- [X] new route types implemented
- [X] existing test suite expanded to cover new route types

## Validation

- [X] tests pass
Screenshot 2024-04-22 at 11 42 05 AM

## ~~Fixes~~ *Partially Addresses*

- #12404

### CODE REVIEW

**Review by adleong — COMMENTED** [review]
I'm not able to build this:

```
$ cargo build --features experimental
   Compiling linkerd-policy-controller-k8s-status v0.1.0 (/workspaces/linkerd2/policy-controller/k8s/status)
error[E0433]: failed to resolve: could not find `GrpcRoute` in `k8s_gateway_api`
   --> policy-controller/k8s/status/src/index.rs:660:41
    |
660 |                 group: k8s_gateway_api::GrpcRoute::group(&()),
    |                                         ^^^^^^^^^
    |                                         |
    |                                         could not find `GrpcRoute` in `k8s_gateway_api`
    |                                         help: a struct with a similar name exists: `TcpRoute`
 ```
    
it seems like the experimental feature is not being applied to the dependency maybe? any ideas?

File: policy-controller/k8s/status/src/routes/http.rs [hunk 1]
```diff
@@ -1,30 +1,5 @@
...
-
-#[derive(Clone, Eq, PartialEq)]
-pub enum BackendReference {
-    Service(ResourceId),
-    Unknown,
-}
+use super::{BackendReference, ParentReference};
+use linkerd_policy_controller_k8s_api::gateway::{CommonRouteSpec, HttpBackendRef};
 
 pub(crate) fn make_parents(namespace: &str, route: &CommonRouteSpec) -> Vec<ParentReference> {
```
Comment (adleong) [inline policy-controller/k8s/status/src/routes/http.rs:4]: Any reason this is defined in http.rs and then re-exported in grpc.rs instead of just being in mod.rs?

File: policy-controller/k8s/status/src/index.rs [hunk 2]
```diff
@@ -51,7 +49,7 @@ pub struct Controller {
     updates: mpsc::Receiver<Update>,
     patch_timeout: Duration,
 
-    /// True if this policy controller is the leader — false otherwise.
```
Comment (adleong) [inline policy-controller/k8s/status/src/index.rs:54]: Lets avoid grammar changes unless they improve clarity.

File: policy-controller/k8s/status/src/index.rs [hunk 3]
```diff
@@ -91,10 +89,11 @@ pub struct IndexMetrics {
 }
 
 #[derive(Clone, PartialEq)]
-struct HttpRoute {
+struct RouteRef {
     parents: Vec<ParentReference>,
     backends: Vec<BackendReference>,
     statuses: Vec<gateway::RouteParentStatus>,
+    r#type: crate::routes::RouteType,
```
Comment (adleong) [inline policy-controller/k8s/status/src/index.rs:96]: I think we can avoid adding this RouteType type altogether. These route refs are stored in a HashMap keyed by NamespaceGroupKindName.  Therefore, we know the kind of the resource from the key.

File: policy-controller/k8s/status/src/lib.rs [hunk 4]
```diff
@@ -1,6 +1,7 @@
-mod http_route;
+#![allow(unused_qualifications)]
```
Comment (adleong) [inline policy-controller/k8s/status/src/lib.rs:1]: why do we want unused_qualifications?

File: policy-controller/k8s/status/src/resource_id.rs [hunk 5]
```diff
@@ -17,3 +18,28 @@ pub struct NamespaceGroupKindName {
     pub namespace: String,
     pub gkn: GroupKindName,
 }
+
+impl<'resource, Resrc> From<&'resource Resrc> for NamespaceGroupKindName
```
Comment (adleong) [inline policy-controller/k8s/status/src/resource_id.rs:22]: I think we can elide the lifetimes here

File: policy-controller/k8s/status/Cargo.toml [hunk 6]
```diff
@@ -25,6 +25,11 @@ thiserror = "1"
...
-version = "1"
-features = ["macros"]
+[dev-dependencies]
+rstest = "^0.19"
+pretty_assertions = "^1.4"
+tokio = { version = "1", features = ["macros"] }
+
+[features]
+default = []
+experimental = ["k8s-gateway-api/experimental"]
```
Comment (adleong) [inline policy-controller/k8s/status/Cargo.toml:35]: I suspect there's not much value in putting this behind a feature.  I doubt that we're going to want the ability to build Linkerd without grpcroute support.  If anything, it would be more useful to gate this behind a runtime flag rather than a build time one.

**Comment by the-wondersmith** [comment]
> I'm not able to build this:
> 
> ```
> $ cargo build --features experimental
>    Compiling linkerd-policy-controller-k8s-status v0.1.0 (/workspaces/linkerd2/policy-controller/k8s/status)
> error[E0433]: failed to resolve: could not find `GrpcRoute` in `k8s_gateway_api`
>    --> policy-controller/k8s/status/src/index.rs:660:41
>     |
> 660 |                 group: k8s_gateway_api::GrpcRoute::group(&()),
>     |                                         ^^^^^^^^^
>     |                                         |
>     |                                         could not find `GrpcRoute` in `k8s_gateway_api`
>     |                                         help: a struct with a similar name exists: `TcpRoute`
> ```
> 
> it seems like the experimental feature is not being applied to the dependency maybe? any ideas?

Ah, it's actually probably that the [PR](https://github.com/linkerd/k8s-gateway-api-rs/pull/94) that adds `GrpcRoute` type to the `k8s-gateway-api` crate hasn't been merged yet 😅.

File: policy-controller/k8s/status/src/index.rs (hunk 2 above)
Comment (the-wondersmith) [reply policy-controller/k8s/status/src/index.rs:54]: Ah, yeah, I've got my IDE set to be almost irritatingly pedantic[¹](#um-actually) and fixing the squigglies is essentially force of habit at this point. I'll revert the change.

<details id="um-actually">
  <summary>visual reference</summary>
  my ide
</details>

File: policy-controller/k8s/status/src/lib.rs (hunk 4 above)
Comment (the-wondersmith) [reply policy-controller/k8s/status/src/lib.rs:1]: There are a few places where I opted for fully qualified paths for certain types even though the parent module was already in scope, to (hopefully) make it a little clearer that even though the types in question are _extremely_ similar (even having the same _name_ in some cases), they're actually two different types / from two different _crates_.

For example, in [`index.rs`](https://github.com/linkerd/linkerd2/blob/main/policy-controller/k8s/status/src/index.rs#L489) it's awfully hard to tell that the `IndexNamespacedResource<k8s::policy::HttpRoute>` block is referring to _linkerd's_ `HttpRoute` type without scrolling all the way back up to [line ~13](https://github.com/linkerd/linkerd2/blob/main/policy-controller/k8s/status/src/index.rs#L13)) to see that `k8s::policy` in this instance is actually `linkerd_policy_controller_k8s_api::policy`. _Granted_ that's not the _best_ example, nor is the argument _particularly_ strong if you're _already_ familiar with the codebase, but once you start add in things like `gateway` actually being `linkerd_policy_controller_k8s_api::gateway` and **_not_** `k8s_gateway_api` into the same file thing can get confusing with _a quickness_. The additional context / clarity certainly helped _me_ keep things straight(er) mentally speaking, so I figured it might also help the next dev who has to pick it up `¯\(ツ)/¯`.

<details>
  <summary>visual aid</summary>
screenshot
</details>

tl;dr the `allow` there just stops clippy from complaining about it. It's not a hill I'm willing to die on though, so if the above doesn't hold water in the team's opinion I'll remove that and the extraneous qualifiers.

File: policy-controller/k8s/status/src/resource_id.rs (hunk 5 above)
Comment (the-wondersmith) [reply policy-controller/k8s/status/src/resource_id.rs]: Fair enough. Change applied.

File: policy-controller/k8s/status/src/tests/routes.rs [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+const TEST_SERVICE_NAME: &str = "srv-8080";
+
+#[rstest]
+#[case::linkerd_http_route(RouteType::LinkerdHttp)]
+#[case::gateway_http_route(RouteType::GatewayHttp)]
+#[cfg_attr(
+    feature = "experimental",
+    case::gateway_grpc_route(RouteType::GatewayGrpc)
+)]
+fn route_accepted_after_server_create(#[case] route_type: RouteType) {
```
Comment (olix0r) [inline policy-controller/k8s/status/src/tests/routes.rs:28]: I'm pretty uncomfortable with taking a new dependency on a new testing framework. I appreciate that rstest can provide some flexibility and code reuse here, but this will leave us in a place where we have one special test written in rstest, and the remainder of our codebase written without such a framework. Should new tests use rstest?

While we've gotten pretty far without using these kinds of tests frameworks, we're not totally opposed to it; but the decision to adopt a testing framework should probably be decoupled from supporting GRPCRoute.

Is it prohibitively difficult to add grpcroute tests without modifying the httproute tests?

**Review by olix0r — COMMENTED** [review]
It looks like something funky has happened with the git history of this branch. Can you merge main or rebase so the diff includes only the relevant changes?

File: policy-controller/k8s/status/Cargo.toml (hunk 6 above)
Comment (the-wondersmith) [reply policy-controller/k8s/status/Cargo.toml:35]: A wire may have gotten crossed somewhere 🤔. AIRC, the `experimental` feature flag was the idea Oliver floated at the end of the original review meeting we had (4/15) to address how we could to do the dev work in smaller, mergable chunks _without_ potentially introducing bugs or breaking changes into released builds.

Catching this sort of thing early _was_ the point of requesting the review though, so..<br/><p align="center">the system works</p>

Assuming we _don't_ want to merge _anything_ into main until the entire feature is complete though, I supposed the alternative would be to make PR's to the main branch of _my_ fork, request reviews from the team against _those_, and once everything is actually ready to go make the PR from my fork to `linkerd/linkerd2`.

Thoughts or suggestions?

File: policy-controller/k8s/status/src/tests/routes.rs (hunk 7 above)
Comment (the-wondersmith) [reply policy-controller/k8s/status/src/tests/routes.rs:28]: FWIW, `rstest` is more than 5 years old [[ref]](https://crates.io/crates/rstest/versions?sort=date#:~:text=OVER%205%20YEARS%20AGO) and is [fairly widely used](https://crates.io/crates/rstest#:~:text=Testing-,Stats%20Overview,-10%2C543%2C541). It is (effectively) to rust what [pytest](https://docs.pytest.org/en/8.1.x/) is to python. I would absolutely argue for the inclusion of `rstest` as standard for new tests, especially considering that doing so would not also introduce any requirement to "convert" existing tests. It just makes writing tests for things with shared behavior much nicer / easier.

Decoupling its adoption from the GRPCRoute support is, of course, totally valid. I just didn't know what side we'd come down on in the face of the inevitable "gotta start somewhere" argument 😅. 

It's not _prohibitively_ (or even _particularly_ difficult) to add grpcroute tests without it, it's just... _ugly_ (for lack of a better word). The tests for each route type _should™_ be the same as they'd all be testing the same behavior, namely "given `{route}`, does the code `{expected behavior}` where only `{route}` would be expected to be different between test cases.

Without some way to parametrize a given test, the test suite becomes repetitive _fast_ which in turn has a real bad tendency to lead to things getting overlooked or missed when future updates are made. These _specific_ tests admittedly aren't the _best_ examples, but suffice it to say that "just add `#[case::amazing_future_type(RouteType::TheRouteToEndAllRoutes)]`" seems like a preferable ideal to shoot for.

For clarity's sake - should I revert the tests to be as they were and re-write them from that point? Or is the decoupling hill not one we're willing to die on?

File: policy-controller/k8s/status/Cargo.toml (hunk 6 above)
Comment (the-wondersmith) [reply policy-controller/k8s/status/Cargo.toml:35]: This was resolved in stand today, adding here for posterity: the strategy here is going to be to make a feature branch off `main`, make and merge PRs for the `GRPCRoute` workstream into _that_ branch, and once everything is deemed ready merge the feature branch into `main`.

**Comment by the-wondersmith** [comment]
> It looks like something funky has happened with the git history of this branch. Can you merge main or rebase so the diff includes only the relevant changes?

@olix0r Yeah, not sure what happened 😅. Rebased onto `main`, the diff should be clean now.

File: policy-controller/k8s/status/src/lib.rs (hunk 4 above)
Comment (adleong) [reply policy-controller/k8s/status/src/lib.rs:1]: I think this can be solved by being consistent about how we refer to these types and potentially NOT adding a `use` for a type directly if we plan to refer to it with a more qualified name.  Using import renames is also a tool we can use to help with clarity e.g. something like `use linkerd_policy_controller_k8s_api::gateway as linkerd_gateyway_api` or similar

**Comment by the-wondersmith** [comment]
In light of the salient points made by Alex (re: passing existing types through instead of adding a new one) and Oliver (re: separating introduction of a new testing paradigm from feature work), along with _whatever_ is going on with the branch history 🤔, I'm opting to close this PR. I'll open a new one once I've refactored.
